### PR TITLE
Implemented escaping

### PR DESCRIPTION
**Pull request by delight-aug**
Wrote custom parser that supports escaping of format directives, resolving #72. This also lowered the needed Rust version to 1.51.0, fixing #78.

Behavioral changes:
- Escaping is needed even if no arguments are provided
- Added debug_assert for when using `{n}` in non-plural gettext invocation

The rule for malformed escaping is that you won't get a singular brace if it's not escaped. See "partial_escaping" unit-test.
I also merged some tests, and reordered tests and definitions of pgettext and npgettext macros, because it was driving me crazy.

### CODE REVIEW

**Comment by Minoru** [comment]
Hey, sorry I haven't reviewed this yet — I'm buried under some IRL stuff and I'm not sure when exactly I'll get to this PR. A week at least, maybe two. Sorry about that!

**Review by Minoru — CHANGES_REQUESTED** [review]
Sorry it took me so long, but I'm finally here :)

The code looks pretty good. There's an issue with imbalanced braces though, see my comment below.

One thing that's missing is docs :P Instead of repeating the same thing for each function, let's have a "Formatting" section at the top of the crate, and link to it from all the docstrings.

<hr/>

A few lower-level comments:

1. in the future, please split commits like https://github.com/Koka/gettext-rs/pull/83/commits/8b098e8f6bfe92749d896baaeb962cf91516e41e in two: one for reordering, another one for actual changes. It was a bit of a pain to untangle the diff;
2. please try to make the commit messages more descriptive. It's nice that the PR itself has a "cover letter" explaining the changes, but someone reading a `git log` would have a hard time trying to quickly grasp what's going on;
3. GitHub is [quite picky](https://docs.github.com/en/github/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests) about its keywords, so "resolving #72" and "fixing #78" don't actually do anything. Change that to "resolves #72", and GitHub will understand you correctly;
4. I don't think this fixes #78, because we still have no idea what Rust version we supported previously. Thus, we don't know if 1.51.0 is good enough, or if we should support earlier versions as well. I suggest we leave that for another PR.

<hr/>

Thank you for working on this, and for waiting for my slowpoke response. You rock! 🤘

File: gettext-rs/src/macros.rs
Comment (delight-aug) [reply gettext-rs/src/macros.rs:249]: Oops, IDK how I haven't thought about adding that debug assert. About the second point — I think it's better to output the least broken string. Having placeholders in place should give users an understanding that something had to be there but somehow isn't. Removing placeholders can break the meaning of the text, or worse — change it. So, I think having placeholders in place is a better decision.

Comment (Minoru) [reply gettext-rs/src/macros.rs:249]: It could be argued that an invalid format string doesn't have meaning, so there's nothing to break :) Also, keeping the brace would be inconsistent with how we handle unsupported specifiers (see #85): I'd like us to consume specifiers, so we should consume any other invalid stuff too. Does that make sense, or are you still not convinced? (This issue is partly a matter of taste, so I'm arguing with you in hopes of finding out which position is more defensible :)

Comment (Minoru) [reply gettext-rs/src/macros.rs:249]: I take it I failed to convince you here, @delight-aug?

Comment (Minoru) [reply gettext-rs/src/macros.rs:249]: Oh, okay, sorry both hurrying you :) I think the draft status is only useful when you first submit a PR, so don't bother with it now. Once you're done editing and CI passes, you can press "Re-request review" next to my nickname in the top right, and I'll get a ping to look at this again.

Thanks for working on this! :+1:

Comment (delight-aug) [reply gettext-rs/src/macros.rs:249]: Actually, I think I misunderstood you a bit. I agree that macros _should_ swallow unsupported and (*may* swallow) unused directives. However, I disagree with your proposition on escaping. In [this playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=7469ede9f3d63c504a440e0b2bfd7631) you might notice that `{{}` and `{}}` give the same error message "unmatched `}` in format string". It's because in the first case `{{}`, it sees an escaped opening brace and an unmatched unescaped closing brace. And in the second one (`{}}`), it's a valid format directive and an unmatched unescaped closing brace. And so, logically, the output should be `{` and `arg`. I think `debug_assert` should be enough to notice the not-entirely-escaped format directive (`{{}`/`{}}`), and making this use-case specific feature is a bit unlogical, probably won't help a lot, and isn't worth of effort. However, the final decision is up to you, @Minoru.

Comment (Minoru) [reply gettext-rs/src/macros.rs:249]: Your examples do make sense, but I feel uneasy about making such an arbitrary decision.

Instead, I looked for some prior art:
* Python's `format` method throws `ValueError` if the format string contains imbalanced curly braces
* C++ `std::vformat` [throws](https://en.cppreference.com/w/cpp/utility/format/vformat) too
* [runtime-fmt](https://crates.io/crates/runtime-fmt) and [dynfmt](https://crates.io/crates/dynfmt) crates return `Result`

I also got thinking about potential users of this feature, and looked through our "dependents" on crates.io:
* GTK apps mentioned in #18. They use Rust-style formats. They assume the format string is correct (i.e. they don't panic and don't return `Result)
* my own app, Newsboat, where I wrap `strnprintf` C function. I use printf-style formats. I assume the format string is correct
* gstore crate [has its own macro](https://gitlab.com/loers/gstore/-/blob/5e367356a7db2254635241224d238fdb0d09570f/src/l10n.rs#L131). It uses Rust-style formats. Assumes the format string is correct
* [tr crate](https://crates.io/crates/tr). Rust-style formats, assumes correctness
* [serde_yaml crate](https://github.com/cecton/serde-gettext) uses Rust-style formats, returns `Result`

I start to think that all this formatting business should be punted onto a separate crate, and gettext-rs should just deal with safe wrappers around gettext functions. Let me sleep on this for a day, I'll get back to you tomorrow. (You can try to persuade me in the meantime :)

Comment (delight-aug) [reply gettext-rs/src/macros.rs:249]: I think functions in the gettext-rs crate can fulfill the purpose of giving a safe API, while macros can be wrappers for formatting them. We could separate macros to their own crate or feature. Also, we'd better support different formatting backends depending on the dependency used in the binary application. This way, we can have all of the formatting features without hassle. 

However, it's unclear what to do with return types. From the perspective of the user of macros, with Result, I could call auto-bug-reporting and display a message to users, that something went wrong. Or I could just unwrap it all the time and let my program crash because of some string. Also, it doesn't look good when you want to include a phrase in GtkLabel, but you need to match it first (or unwrap it in place). Maybe a good decision is to let users specify an action to run on a broken string? So they set up a closure near textdomain bindings. And by default just swallow things... Just spamming ideas

Comment (Minoru) [reply gettext-rs/src/macros.rs:249]: Yeah, my tentative plan is to remove our macros and open an issue explaining what we tried and what problems we encountered. Interested users would comment with their thoughts, perhaps we'll find a new design, which we will then implement. In the meantime, anyone can create new crates on top of gettext-rs to try out ideas.

I'll be sure to include your ideas in that issue :)

**Comment by Minoru** [comment]
Okay, I filed #86, so let me close this. You've put a lot of work into macros — thanks for that! I'm sorry this has to be shelved for now. Perhaps we'd be able to make a better job when we got a bit more input on what people want.

**Comment by delight-aug** [comment]
Sure! Thank you for being nice and forcing me to use git properly!

**Pull request closed without merging**
